File: src/main/java/algo/_02linkedlist/SinglyLinkedList.java
package algo._02linkedlist;

import common.ListNode;

public class SinglyLinkedList {

    private ListNode head = null;

    public SinglyLinkedList (int[] data) {
        if (data == null || data.length == 0) return;

        for (int i = 0; i < data.length; i++) {
            if (i == 0) {
                head = new ListNode(data[i]);
            } else {
                insertToTail(data[i]);
            }
        }
    }

    public SinglyLinkedList () {

    }

    public ListNode findByValue (int value) {
        ListNode p = head;

        while (p != null && p.getValue() != value) {
            p = p.getNext();
        }

        return p;
    }

    public ListNode findByIndex (int index) {
        ListNode p = head;
        int pos = 0;
        while (p != null && pos != index) {
            p = p.getNext();
            pos++;
        }

        return p;
    }

    /**
     * 在链表头部插入值为value的节点
     *
     * @param value
     */
    public void insertToHead (int value) {
        ListNode newHead = new ListNode(value);
        insertToHead(newHead);
    }

    /**
     * 在链表头部插入节点newNode
     *
     * @param newNode
     */
    public void insertToHead (ListNode newNode) {
        if (head == null) {
            newNode.setNext(null);
        } else {
            newNode.setNext(head);
        }
        head = newNode;
    }

    /**
     * 在链表尾部插入值为value的节点
     *
     * @param value
     */
    public void insertToTail (int value) {
        ListNode newNode = new ListNode(value);
        insertToTail(newNode);
    }

    /**
     * 在链表尾部插入节点newNode
     *
     * @param newNode
     */
    public void insertToTail (ListNode newNode) {
        if (head == null) {
            head = newNode;
        } else {
            ListNode q = head;
            while (q.getNext() != null) {
                q = q.getNext();
            }
            newNode.setNext(q.getNext());
            q.setNext(newNode);
        }
    }


    /**
     * 在节点P后插入值为value的节点
     *
     * @param p
     * @param value
     */
    public void insertAfter (ListNode p, int value) {
        ListNode newNode = new ListNode(value);
        insertAfter(p, newNode);
    }

    /**
     * 在节点P后插入新节点newNode
     *
     * @param p
     * @param newNode
     */
    public void insertAfter (ListNode p, ListNode newNode) {
        if (p == null) return;

        newNode.setNext(p.getNext());
        p.setNext(newNode);
    }

    /**
     * 在节点p之前插入值为value的节点
     *
     * @param p
     * @param value
     */
    public void insertBefore (ListNode p, int value) {
        ListNode newNode = new ListNode(value);
        insertBefore(p, newNode);
    }

    /**
     * 在节点p之前插入节点newNode
     *
     * @param p
     * @param newNode
     */
    public void insertBefore (ListNode p, ListNode newNode) {
        if (head == null) return;

        if (p == head) {
            insertToHead(newNode);
            return;
        }

        ListNode q = head;
        while (q != null && p != q.getNext()) {
            q = q.getNext();
        }

        if (q == null) {
            return;
        }

        newNode.setNext(p);
        q.setNext(newNode);
    }

    /**
     * 删除链表中**第一个**值为value的节点
     *
     * @param value
     */
    public void deleteFirstByValue (int value) {
        if (head == null) return;

        ListNode p = head;
        ListNode q = null;

        while (p.getNext() != null && p.getValue() != value) {
            q = p;
            p = p.getNext();
        }

        if (p == null) return;

        if (q == null) {
            head = head.getNext();
        } else {
            q.setNext(p.getNext());
        }

    }

    /**
     * 删除链表中**所有**值为value的节点
     *
     * @param value
     */
    public void deleteAllByValue (int value) {

        if (head == null) return;

        ListNode p = head;
        ListNode q = null;
        while (p != null) {
            if (p.getValue() == value) {
                if (p == head) {
                    head = head.getNext();
                    p = head;
                } else {
                    if (q != null) {
                        q.setNext(p.getNext());
                        p = p.getNext();
                    }
                }
                continue;
            }
            q = p;
            p = p.getNext();
        }
    }

    public void printLinkedList () {
        if (head == null) System.out.println("linked list is null");

        String content = "linked list: \n";
        ListNode p = head;
        while (p != null) {
            content += p.getValue() + "\n";
            p = p.getNext();
        }

        System.out.println(content);
    }

    /**
     * 删除链表中值为value的节点
     *
     * @param value
     * @param delAll true: 删除所有值为value的节点 false: 删除第一个值为value的节点
     */
    public void deleteByValue (int value, boolean delAll) {

        if (delAll) {
            deleteAllByValue(value);
        } else {
            deleteFirstByValue(value);
        }
    }

    /**
     * 删除节点delNode
     *
     * @param delNode
     */
    public void deleteByNode (ListNode delNode) {
        if (delNode == null || head == null) return;

        if (head == delNode) {
            head = head.getNext();
            return;
        }

        ListNode p = head;
        while (p.getNext() != null) {
            if (p.getNext() == delNode) {
                p.setNext(p.getNext().getNext());
                break;
            }
            p = p.getNext();
        }
    }

    /**
     * 判断两个链表数据是否完全一致
     *
     * @param left
     * @param right
     * @return
     */
    public boolean TFResult (ListNode left, ListNode right) {
        ListNode l = left;
        ListNode r = right;

        while (l != null && right != null) {
            if (l.getValue() == right.getValue()) {
                l = l.getNext();
                r = r.getNext();
                continue;
            } else {
                break;
            }
        }

        if (l == null && r == null) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * 链表中环的检测
     * 空间复杂度O(1)
     * 时间复杂度O(n)
     */
    public boolean hadCycle (ListNode head) {
        if (head == null) return false;

        ListNode fast = head.getNext();
        ListNode slow = head;

        /**
         * 思路很厉害，从数学上讲：
         * 有环的链表，快指针一定会绕回来追赶慢指针，
         * 相对速度快1，所以一定可以重合
         */
        while (fast != null && fast.getNext() != null) {
            fast = fast.getNext().getNext();
            slow = slow.getNext();

            if (slow == fast) return true;
        }

        return false;
    }

    /**
     * 链表中环的检测
     * 返回第一个成环点
     * @param head
     * @return
     */
    public ListNode detectCycle (ListNode head) {
        if (head == null || head.getNext() == null || head.getNext().getNext() == null) return null;

        ListNode fast = head.getNext().getNext();
        ListNode slow = head.getNext();

        //找到相遇点
        while (fast != slow) {
            if (fast.getNext() != null && fast.getNext().getNext() != null) {
                fast = fast.getNext().getNext();
                        slow = slow.getNext();
            } else {
                return null;
            }
        }
        //找到入口
        fast = head;
        while (fast != slow) {
            fast = fast.getNext();
            slow = slow.getNext();
        }

        return slow;
    }


    /**
     * 判断回文数
     *
     * @return
     */
    public boolean isPalindromicNumber () {

        if (head == null) return false;

        if (head.getNext() == null) {
            System.out.println("只有一个节点");
            return true;
        }

        ListNode fast = head;
        ListNode slow = head;
        SinglyLinkedList list = new SinglyLinkedList();
        list.insertToHead(head);

        while (fast.getNext() != null && fast.getNext().getNext() != null) {
            fast = fast.getNext().getNext();
            slow = slow.getNext();
            list.insertToHead(slow);
        }

        ListNode left = null;
        ListNode right = null;
        if (fast.getNext() == null) {
            //整个队列节点数为奇数
            //slow是中点
            left = list.head;
            right = slow;
        } else {
            //整个队列节点数为偶数
            //slow 和 slow.next是中点
            left = list.head;
            right = slow.getNext();
        }

        return TFResult(left, right);
    }

}

File: kotlin/src/_03linkedlist/SinglyLinkedList.kt
package _03linkedlist

import utils.Node

class SinglyLinkedList {

    var head: Node? = null

    constructor(array: IntArray) {

        for (v in array) {
            insertToTail(v)
        }
    }

    constructor(array: Array<Int?>) {
        for (v in array) {
            insertToTail(v!!)
        }
    }

    /**
     * 在链表尾部插入值为value的节点
     *
     * @param value
     */
    fun insertToTail(value: Int) {
        val newNode = Node(value)
        insertToTail(newNode)
    }

    /**
     * 在链表尾部插入节点newNode
     *
     * @param newNode
     */
    fun insertToTail(newNode: Node) {

        if (head == null) {
            head = newNode
        } else {
            var q = head

            while (q?.next != null) {
                q = q.next
            }
            newNode.next = q?.next
            q?.next = newNode
        }
    }

    fun print () {

        var content = ""
        var p = head
        while (p != null) {
            content += "" + p.value + "\n"
            p = p.next
        }
        println(content)
    }

    /**
     * 单链表反转
     * 不带头
     *
     * 时间复杂度: O(n)
     * 控件复杂度: O(1)
     */
    fun reverseList (head: Node?) : Node? {

        var newList: Node? = null
        var p = head
        var pNext: Node?

        while (p?.next != null) {
            pNext = p.next

            if (newList == null) {
                p.next = null
                newList = p
            } else {
                p.next = newList
                newList = p
            }

            p = pNext
        }

        p?.next = newList
        newList = p

        return newList
    }

    /**
     * 单链表反转
     * 带头
     */
    fun revereListWithHead (p: Node?) : Node? {

        //指定一个head
        var head = Node(8080)

        //head永远是head，遍历的买一个新节点插在head后面
        var q = p
        var qNext: Node? = null
        while (q != null) {
            qNext = q.next

            q.next = head.next
            head.next = q

            q = qNext
        }

        return head

    }

    /**
     * 链表中环的检测
     */
    fun hadCycle (head: Node?): Boolean {
        if (head == null) return false

        var fast = head.next
        var slow = head

        /**
         * 思路很厉害，从数学上讲：
         * 有环的链表，快指针一定会绕回来追赶慢指针，
         * 相对速度快1，所以一定可以重合相对速度快1，所以一定可以重合
         */
        while (fast != null && fast.next != null) {
            fast = fast?.next?.next
            slow = slow?.next

            if (slow == fast) return true
        }

        return false
    }

    /**
     * 链表中的换检测：开始入环的第一个节点
     */
    fun detectCycle (head: Node?) : Node? {
        if (head == null || head.next == null || head.next?.next == null) return null

        var fast = head.next?.next
        var slow = head.next

        //找到相遇点
        while (fast != slow) {
            if (fast?.next != null && fast?.next?.next != null) {
                fast = fast.next?.next
                slow = slow?.next
            } else {
                return null
            }
        }
        //找到入口
        fast = head
        while (fast != slow) {
            fast = fast?.next
            slow = slow?.next
        }

        return slow
    }

    /**
     * 合并两个有序链表
     *
     * 直观的做法(未使用哨兵节点)
     */
//    fun mergeTwoOrderedList (l1: Node?, l2: Node?) : Node? {
//
//        if (l1 == null) {
//            return l2
//        } else if (l2 == null) {
//            return l1
//        }
//
//        var head = l1
//        var p1 = head
//        var p2 = l2
//        var p1Pre: Node? = null
//        var p2Next: Node? = null
//
//        //把p2并入p1
//        while (p1 != null && p2 != null) {
//
//            //这里不可能是head
//            if (p1.value < p2!!.value) {
//
//                if (p1.next!!.value >= p2.value) {
//                    //在p1 与 p1.next之间：插入
//                    p2Next = p2.next
//
//                    p2.next = p1.next
//                    p1.next = p2
//
//                    p2 = p2Next
//                } else {
//                    //在p1右边，p1右移，再次运算
//                    p1Pre = p1
//                    p1 = p1.next
//                }
//            } else {
//
//                //p1是头结点，直接插入
//                if (p1 == head) {
//                        p2Next = p2.next
//                        p2.next = p1
//                        head = p2
//                        p1Pre = p2
//                        p2 = p2Next
//                } else {
//                    if (p1Pre!!.value <= p2!!.value) {
//                        //p2在p1 与 p1Pre之间
//                        p2Next = p2.next
//
//                        p2.next = p1
//                        p1Pre.next = p2
//                        p1Pre = p2
//
//                        p2 = p2Next
//                    } else {
//                        //p2在p1Pre之前，有序列表中，这种情况不会出现
//                    }
//                }
//            }
//        }
//
//        if (p2 != null) {
//            p1Pre?.next = p2
//        }
//
//        //使用原来的空间
//        return head
//    }

    /**
     * 合并两个有序列表
     * 使用了soldier节点简化问题
     */
    fun mergeTwoOrderedList (l1: Node?, l2: Node?) : Node? {

        var soldier = Node(0)
        var p = soldier
        var p1 = l1
        var p2 = l2


        while (p1 != null && p2 != null) {
            if (p1.value < p2.value) {
                p.next = p1
                p1 = p1.next
            } else {
                p.next = p2
                p2 = p2.next
            }
            p = p.next!!
        }

        if (p1 == null) p.next = p2
        if (p2 == null) p.next = p1

        return soldier.next
    }

    /**
     * 删除链表中的倒数第N个节点
     *
     */
    fun removeNthFromEnd(head: Node?, n: Int): Node? {
        //思路：k已知，指针指向length-k+1个位置即可
        var fast = head
        var pos = 1

        while (fast != null && pos < n) {
            fast = fast.next
            ++pos
        }

        if (fast == null) return head

        var slow = head
        var prev: Node? = null

        while (fast?.next != null) {

            fast = fast?.next
            prev = slow
            slow = slow?.next
        }

        if (prev == null) {
            return head?.next
        } else {
            prev.next = prev.next?.next
            return head
        }

    }

    fun findMidNode (head: Node?) : Node? {

        if (head == null) return null

        var fast = head
        var slow = head

        if (fast.next != null && fast.next?.next != null) {
            fast = fast.next?.next
            slow = slow.next
        }

        return slow
    }

    companion object {

        fun printList (head: Node?) {

            var content = ""
            var p = head
            while (p != null) {
                content += "" + p.value + "\n"
                p = p.next
            }
            println(content)
        }
    }

}